aignostic/aignostic/dataset/validate_dataset_api.py

```
@@ -1,65 +1,63 @@
-    """
-    Validate a dataset URL and return the columns and number of rows.
+
+@app.post("/fetch-data")
+def fetch_dataset(request: FetchDatasetRequest) -> ModelInput:
     """
-    if not is_valid_url(url):
-        raise HTTPException(status_code=400, detail="Invalid URL")
+    Validate a dataset URL and parse the data to be used in the model.
 
+    Params:
+    - request : FetchDatasetRequest - Pydantic model for the request
+    """
     try:
-        data = fetch_data(url)
-        dataset = parse_dataset(data)
-        return JSONResponse(content={
-            "message": "Data successfully parsed into DataFrame",
-            "columns": dataset.column_names,
-            "rows": len(dataset.rows)
-        })
+        headers = {"Authorization": f"Bearer {request.dataset_api_key}"} if request.dataset_api_key else {}
+        response = requests.get(request.dataset_url, headers=headers)
+
+        response.raise_for_status()
 
-    # TODO: Add logging to exception handling
+        data = response.json()
+        return validate_dataset_format(data)
     except requests.exceptions.RequestException as e:
-        if isinstance(e, requests.exceptions.HTTPError) and e.response.status_code == 404:
-            raise HTTPException(status_code=404, detail="Resource not found")
-        else:
-            raise HTTPException(status_code=400, detail="Failed to fetch data")
-    except Exception:
-        raise HTTPException(status_code=400, detail="Invalid data format")
+        if response.status_code == 401:
+            raise HTTPException(status_code=401, detail="Unauthorized access: Please check your API Key")
+        elif response.status_code == 404:
+            raise HTTPException(status_code=404, detail="Data not found")
+        raise HTTPException(status_code=response.status_code, detail=f"Error while fetching data: {e}")
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=f"Error while validating data: {e}")
```

Separating into a more specific try-except block just for calling the validate_dataset_format function would be a better, more fine-grained solution. Though in fact, I believe this exception shouldn't be caught at all. 

Since we need to raise a HTTPException anyways to propagate to our front end, it would be simpler to raise HTTPExceptions instead of ValueErrors and just let them propagate to the frontend without being caught (they are caught in the frontend). 

In my opinion nesting the error message as "Error while validating data: {e}" makes the error message longer to read and can be annoying (in my opinion) but would love to know what you think.

---

Smart, I have separated the try-except block for `validate_dataset_format`. Thanks for your error-handling guidance.